src/plugins/navigation_embeddable/public/components/external_link/external_link_destination_picker.tsx

```diff
@@ -6,31 +6,66 @@
  * Side Public License, v 1.
  */
 
+import React, { useCallback, useState } from 'react';
 import useMount from 'react-use/lib/useMount';
-import React, { useState } from 'react';
 
 import { EuiFieldText } from '@elastic/eui';
+import { urlDrilldownValidateUrl } from '@kbn/ui-actions-enhanced-plugin/public';
+
 import { ExternalLinkStrings } from './external_link_strings';
+import { coreServices } from '../../services/kibana_services';
 
-// TODO: As part of https://github.com/elastic/kibana/issues/154381, replace this regex URL check with more robust url validation
-const isValidUrl =
-  /^https?:\/\/(?:www.)?[-a-zA-Z0-9@:%._+~#=]{1,256}.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_+.~#?&/=]*)$/;
+class DisallowedUrlError extends Error {}
 
 export const ExternalLinkDestinationPicker = ({
   onDestinationPicked,
+  setDestinationError,
   initialSelection,
   ...other
 }: {
   onDestinationPicked: (destination?: string) => void;
+  setDestinationError: (error: string | undefined) => void;
   initialSelection?: string;
 }) => {
   const [validUrl, setValidUrl] = useState<boolean>(true);
   const [currentUrl, setCurrentUrl] = useState<string>(initialSelection ?? '');
 
+  const validateUrl = useCallback(
+    (url: string) => {
+      try {
+        const allowedUrl = coreServices.http.externalUrl.validateUrl(url);
```

Comment:
This tests to ensure that the URL is allowed according to the `externalUrl.policy` configuration